iml-api/src/graphql.rs

```diff
@@ -143,6 +144,94 @@ impl QueryRoot {
 
         Ok(xs)
     }
+    #[graphql(arguments(
+        limit(description = "paging limit, defaults to 20"),
+        offset(description = "Offset into items, defaults to 0"),
+        dir(description = "Sort direction, defaults to asc"),
+        args(description = "Snapshot listing arguments")
+    ))]
+    /// Fetch the list of snapshots
+    async fn snapshots(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        dir: Option<SortDir>,
+        args: List,
+    ) -> juniper::FieldResult<Vec<Snapshot>> {
+        let dir = dir.unwrap_or_default();
+
+        if args.detail {
+            let xs = sqlx::query!(
+                r#"
+                    SELECT filesystem_name, snapshot_name, create_time, modify_time, snapshot_fsname, mounted, comment from snapshot s
+                    WHERE filesystem_name = $4 AND $5::text IS NULL OR snapshot_name = $5
+                    ORDER BY
+                        CASE WHEN $3 = 'asc' THEN s.snapshot_name END ASC,
+                        CASE WHEN $3 = 'desc' THEN s.snapshot_name END DESC
+                    OFFSET $1 LIMIT $2"#,
+                offset.unwrap_or(0) as i64,
+                limit.unwrap_or(20) as i64,
+                dir.deref(),
+                args.fsname,
+                args.name,
+            )
+            .fetch_all(&context.client)
+            .await?;
+
+            let snapshots: Vec<_> = xs
+                .into_iter()
+                .map(|x| Snapshot {
+                    snapshot_name: x.snapshot_name,
+                    filesystem_name: x.filesystem_name,
+                    details: vec![Detail {
+                        comment: x.comment,
+                        create_time: x.create_time,
+                        modify_time: x.modify_time,
+                        snapshot_fsname: x.snapshot_fsname,
+                        // FIXME
+                        snapshot_role: None,
+                        status: x.mounted.map(|b| {
+                            if b {
+                                Status::Mounted
+                            } else {
+                                Status::NotMounted
+                            }
+                        }),
+                    }],
+                })
+                .collect();
+
+            Ok(snapshots)
+        } else {
+            let xs = sqlx::query!(
+                r#"
+                    SELECT filesystem_name, snapshot_name from snapshot s
+                    WHERE filesystem_name = $4 AND $5::text IS NULL OR snapshot_name = $5
+                    ORDER BY
+                        CASE WHEN $3 = 'asc' THEN s.snapshot_name END ASC,
+                        CASE WHEN $3 = 'desc' THEN s.snapshot_name END DESC
```

Comment:
Maybe ordering by creation time would make more sense?

Reply:
I think it would be more beneficial to order by time, personally. If the names are ordered then it's more of a sort by alphabetical name but there is no relative meaning behind the sort. If the snapshots are ordered by time, it could help determine what snapshot to restore. For example, if I had 30 snapshots and I want to quickly go back to the first snapshot when I setu p my system, I would list out the snapshots and scroll to the first one and immediately know that this is the snapshot I want to use.

Reply:
I think the pertinent usecase will be to delete the oldest snapshots so I agree that sorting by time is useful.

Reply:
Done